### PR TITLE
TF-2628 Create open PDF in new browser tab feature

### PR DESCRIPTION
**Pull request by tddang-linagora**
### **Issue**

https://github.com/linagora/tmail-flutter/issues/2628
### **Resolved**

As discussed with @hoangdat , @dab246 & @hieutbui , attachment button for PDF in web version will have 2 functionalities. When user tap on download icon inside the button, openDownloadedFileWeb will be triggered, showing the PDF inside new browser tab and user can download the file here themselves. When tapping anywhere else inside the attachment button, the PDF will be downloaded automatically, and the auto opening feature will be handle by the browser itself depends on its own settings.

- Firefox

https://github.com/linagora/tmail-flutter/assets/160106668/9947e64d-b9db-4670-8d9b-ed6e2810a83d

https://github.com/linagora/tmail-flutter/assets/160106668/27166523-d924-4ca1-8e18-c53867f3f749

- Safari

https://github.com/linagora/tmail-flutter/assets/160106668/8448ef39-46a5-4c6a-a591-fc6efa65eaa8

https://github.com/linagora/tmail-flutter/assets/160106668/c7c72340-0f8d-4ed9-b25e-4c5b0a3a200d

- Chrome

https://github.com/linagora/tmail-flutter/assets/160106668/bc002736-3a56-4d9e-a90e-53fb307a1785

https://github.com/linagora/tmail-flutter/assets/160106668/046ee15e-ffca-44ae-8b58-7abd4f1f341e

### CODE REVIEW

File: docs/adr/0039-logic-for-opening-pdf-in-new-browser-tab.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+# 38. Logic for displaying PDF in new browser tab
```
Comment (dab246) [inline docs/adr/0039-logic-for-opening-pdf-in-new-browser-tab.md:1]:
```suggestion
# 39. Logic for displaying PDF in new browser tab
```

Comment (tddang-linagora) [reply docs/adr/0039-logic-for-opening-pdf-in-new-browser-tab.md]: Done

File: lib/features/email/presentation/controller/single_email_controller.dart [hunk 2]
```diff
@@ -753,6 +785,24 @@ class SingleEmailController extends BaseController with AppLoaderMixin {
...
+    if (mimeType != Constant.pdfMimeType) {
+      _downloadAttachmentForWebSuccessAction(success);
+      return;
+    }
+
+    mailboxDashBoardController.deleteDownloadTask(success.taskId);
+
+    _downloadManager.openDownloadedFileWeb(
+        success.bytes, 
+        success.attachment.type?.mimeType);
```
Comment (dab246) [inline lib/features/email/presentation/controller/single_email_controller.dart:803]: Why not separate actions for `view` and `download` right from clicking on the attachment item in the email view screen. As far as I remember, every attachment that appears in the email view already has a `mimeType`.

File: test/features/email/domain/usecases/view_attachment_for_web_interactor_test.dart [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+  final tTaskId = DownloadTaskId('123');
+  final tAttachment = Attachment();
+  final tAccountId = AccountId(Id('id'));
+  final tException = TimeoutException('321');
+  const tBaseDownloadUrl = 'base_download_url';
+  final tReceiveController = StreamController<Either<Failure, Success>>();
+  final tFailure = MockFailure();
+  final tSuccess = MockSuccess();
+  final tBytes = Uint8List(12);
```
Comment (dab246) [inline test/features/email/domain/usecases/view_attachment_for_web_interactor_test.dart:40]: Please pay attention to the naming. It needs to be more clear.

File: lib/features/email/domain/usecases/view_attachment_for_web_interactor.dart [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+                    success.taskId,
+                    success.attachment,
+                    success.bytes,
+                  ));
+                }
+
+                return right(success);
+              },
+            ),
+          );
```
[the remark this answers is not in the document]
Comment (tddang-linagora) [reply lib/features/email/domain/usecases/view_attachment_for_web_interactor.dart:69]: > Please use {} when handling multiple tasks within the function.
The `map` only has `return` task, so arrow function is ok

File: test/features/email/domain/usecases/view_attachment_for_web_interactor_test.dart (hunk 3 above)
Comment (tddang-linagora) [reply test/features/email/domain/usecases/view_attachment_for_web_interactor_test.dart]: Done

File: lib/features/email/presentation/controller/single_email_controller.dart [hunk 5]
```diff
@@ -744,6 +753,29 @@ class SingleEmailController extends BaseController with AppLoaderMixin {
...
+    final session = mailboxDashBoardController.sessionCurrent;
+    if (accountId != null && session != null) {
+      final baseDownloadUrl = session.getDownloadUrl(jmapUrl: dynamicUrlInterceptors.jmapUrl);
+      final generateTaskId = DownloadTaskId(uuid.v4());
+      consumeState(_viewAttachmentForWebInteractor.execute(
+          generateTaskId,
+          attachment,
+          accountId,
+          baseDownloadUrl,
+          _downloadProgressStateController));
```
[the remark this answers is not in the document]
Comment (tddang-linagora) [reply lib/features/email/presentation/controller/single_email_controller.dart]: Done

**Review by hoangdat — COMMENTED** [review]
no, one commit for all stuffs :))

File: lib/features/email/domain/usecases/view_attachment_for_web_interactor.dart [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+  Stream<Either<Failure, Success>> execute(
+    DownloadTaskId taskId,
+    Attachment attachment,
+    AccountId accountId,
+    String baseDownloadUrl,
+    StreamController<Either<Failure, Success>> onReceiveController,
+  ) =>
+      _downloadAttachmentForWebInteractor
+          .execute(
```
Comment (hoangdat) [inline lib/features/email/domain/usecases/view_attachment_for_web_interactor.dart:26]: what happen if `execute` throw exception?

Comment (tddang-linagora) [reply lib/features/email/domain/usecases/view_attachment_for_web_interactor.dart:26]: The `catch` block inside `DownloadAttachmentForWebInteractor` will handle the exception, so there's no need for another layer of `try`-`catch` in `ViewAttachmentForWebInteractor`

File: test/features/email/presentation/controller/single_email_controller_test.dart [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+  final dynamicUrlInterceptors = MockDynamicUrlInterceptors();
+  final deleteCredentialInteractor = MockDeleteCredentialInteractor();
+  final logoutOidcInteractor = MockLogoutOidcInteractor();
+  final deleteAuthorityOidcInteractor = MockDeleteAuthorityOidcInteractor();
+  final appToast = MockAppToast();
+  final imagePaths = MockImagePaths();
+  final responsiveUtils = MockResponsiveUtils();
+  final uuid = MockUuid();
+
+  late SingleEmailController tController = SingleEmailController(
```
Comment (hoangdat) [inline test/features/email/presentation/controller/single_email_controller_test.dart:103]:
```suggestion
  late SingleEmailController singleEmailController = SingleEmailController(
```

Comment (hoangdat) [reply test/features/email/presentation/controller/single_email_controller_test.dart:103]: Please use explicit variable name. No need `t`, `g`, .... prefix.

Comment (tddang-linagora) [reply test/features/email/presentation/controller/single_email_controller_test.dart]: Done

File: lib/features/email/presentation/widgets/attachment_item_widget.dart [hunk 8]
```diff
@@ -25,14 +27,15 @@ class AttachmentItemWidget extends StatelessWidget {
...
+    this.viewAttachmentAction,
   }) : super(key: key);
 
   @override
   Widget build(BuildContext context) {
     return Material(
       color: Colors.transparent,
       child: InkWell(
-        onTap: () => downloadAttachmentAction?.call(attachment),
+        onTap: () => viewAttachmentAction?.call(attachment),
```
Comment (dab246) [inline lib/features/email/presentation/widgets/attachment_item_widget.dart:38]: So, where is `downloadAttachmentAction` used?

Comment (tddang-linagora) [reply lib/features/email/presentation/widgets/attachment_item_widget.dart:38]: Line 90 same file

File: lib/features/email/presentation/controller/single_email_controller.dart [hunk 9]
```diff
@@ -744,6 +753,29 @@ class SingleEmailController extends BaseController with AppLoaderMixin {
     }
   }
 
+  void viewAttachmentForWeb(Attachment attachment) {
+    _viewAttachmentForWebAction(attachment);
+  }
+
+  void _viewAttachmentForWebAction(Attachment attachment) async {
```
Comment (dab246) [inline lib/features/email/presentation/controller/single_email_controller.dart:760]: IMO, no need `async` at here

Comment (tddang-linagora) [reply lib/features/email/presentation/controller/single_email_controller.dart]: Done

[the remark this answers is not in the document]
Comment (tddang-linagora) [reply lib/features/email/presentation/controller/single_email_controller.dart]: Done

**Review by dab246 — APPROVED** [review]
LGTM

**Review by hoangdat — APPROVED** [review]
please rebase

**Pull request merged**